### PR TITLE
feat: improve type coercion and validation for datafusion execution

### PR DESCRIPTION
**Pull request by silver-ymz**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://raw.githubusercontent.com/risingwavelabs/risingwave/17af8a747593ebdbfa826691daf75bdab7d14fa0/.github/contributor-license-agreement.txt).

## What's changed and what's your intention?

This PR tightens type handling and improves correctness when executing scalar functions and expressions via DataFusion.

### Key changes

1. Insert explicit type casts before evaluating scalar functions
Ensures inputs conform to expected types and avoids relying on implicit or incomplete coercion.

2. Strengthen validation in convert_function_call
    - Enforce stricter checks on DataFusion native types to prevent unsupported or ambiguous conversions.
    - Remove `RegexMatch`, as DataFusion currently lacks full feature support for it.
    - Use `BinaryTypeCoercer` to explicitly validate whether a given binary operation is supported by DataFusion.

3. Enable DataFusion’s analyzer in `execute_datafusion_plan`
This allows DataFusion to apply its own type coercion and analysis logic, improving compatibility and reducing manual edge-case handling.

## Checklist

- [x] I have written necessary rustdoc comments.
- [ ]  I have added necessary unit tests and integration tests.
- [ ]  I have added test labels as necessary. 
- [ ]  I have added fuzzing tests or opened an issue to track them. 
- [ ]  My PR contains breaking changes. 
- [ ]  My PR changes performance-critical code, so I will run (micro) benchmarks and present the results. 
- [ ]  I have checked the [Release Timeline](https://github.com/risingwavelabs/rw-commits-history/blob/main/release_timeline.md) and [Currently Supported Versions](https://docs.risingwave.com/changelog/release-support-policy#support-end-dates-for-recent-releases) to determine which release branches I need to cherry-pick this PR into. 

## Documentation

- [ ]  My PR needs documentation updates.

### CODE REVIEW

**Comment by silver-ymz** [comment]
* **#24120** <a href="https://app.graphite.com/github/pr/risingwavelabs/risingwave/24120?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/risingwavelabs/risingwave/24120?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#24097** <a href="https://app.graphite.com/github/pr/risingwavelabs/risingwave/24097?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24065** <a href="https://app.graphite.com/github/pr/risingwavelabs/risingwave/24065?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/frontend/src/datafusion/convert.rs [hunk 1]
```diff
@@ -90,24 +94,85 @@ pub trait ColumnTrait {
     }
     fn len(&self) -> usize;
     fn column(&self, index: usize) -> Column;
+    fn rw_data_type(&self, index: usize) -> RwDataType;
+    fn df_data_type(&self, index: usize) -> DFDataType;
 }
 
-impl ColumnTrait for DFSchema {
+impl<'a> ColumnTrait for (&'a DFSchema, &'a RwSchema) {
```
Comment (chenzl25) [inline src/frontend/src/datafusion/convert.rs]: How about defining a struct instead of using unnamed tuple? It is easier to identify the usage.
```
pub struct InputColumns<'a> {
    df_schema: &'a DFSchema,
    rw_schema: &'a RwSchema,
}
```
